Add negative tests to increase code coverage

nniranjhana opened the pull request:
Added posix_checks, which does three things:

(i) checks which of the scheduling priorities are supported
and lists the max and min priorities if supported,
(ii) checks which clocks are supported,
(iii) checks if timer APIs with both supported and
unsupported clocks to see if they pass or fail as expected.

In some cases pthread attrib initialization may fail as
garbage value may indicate that this attribute is already
initialized. Hence adding a destroy before initializing it.

Similar to #6499

nniranjhana commented:
Closing this PR, since @youvedeep has merged my changes into his PR #6499 since we do not have time to wait (this if for 1.11), and we modify same files.

nniranjhana commented:
Re-opening, since #6499 is merged. Will update after resolving conflicts with #6499.

Commit:
tests: posix: Fixing pthread attr object initialization.

In some cases pthread attrib initialization may fail as
garbare value may indicate that this attribute is already
initialized. Destroy the attr object and recreate it.

Signed-off-by: Niranjhana N <niranjhana.n@intel.com>

lpereira commented on tests/posix/clock/src/posix_clock.c:
```diff
@@ -11,11 +11,17 @@
 
 	printk("POSIX clock APIs\n");
-	clock_gettime(CLOCK_MONOTONIC, &ts);
+	ret = clock_gettime(CLOCK_REALTIME, &ts);
```
Shouldn't this be two separate tests, one for CLOCK_REALTIME, and another for CLOCK_MONOTONIC?

nniranjhana replied:
Hi @lpereira  This is done exclusively for getting code coverage. We do not support CLOCK_REALTIME so it will return with error (non 0 value). This will help to test condition if wrong argument has been passed with clock_gettime.

lpereira commented on tests/posix/pthread_equal/src/pthread_equal.c:
```diff
@@ -22,12 +22,19 @@ void *thread_top(void *p1)
-	pthread_attr_init(&attr);
+	ret = pthread_attr_init(&attr);
+	if (ret != 0) {
+		zassert_false(pthread_attr_destroy(&attr),
```
Is there any point in calling pthread_attr_destroy() if ret is not 0? pthread_attr_destroy() can be called regardless of success from pthread_attr_init(); otherwise, it's going to be called with a potentially invalid pthread_attr_t struct (won't crash but won't signal an error either).

nniranjhana replied:
Hi @lpereira , 

pthread_attr_init is going to fail only on one condition that - attrib object is already initialized. If attrib object is initialized then it need to be destroyed. Because of that only we are calling pthread_attr_destroy and initializing the attrib object again.
as attrib->status is just a variable and it may have garbage value, so this garbage value may indicate that this attr object is initialized. So to handle those cases I have put this check.

Commit:
tests: posix: add more tests

Added posix_checks, which checks supported priorities,
clocks and timer APIs, mainly in the point of negative
tests to increase code coverage.

Signed-off-by: Niranjhana N <niranjhana.n@intel.com>

pfalcon reviewed (approved):
Given that these are tests and pass, let me give +1.
